keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1086 @@
+  - [Flexible Extension Support](#flexible-extension-support)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+  - [Audit Annotations](#audit-annotations)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Optional Future Extensions](#optional-future-extensions)
+  - [Rollout of baseline-by-default for unlabeled namespaces](#rollout-of-baseline-by-default-for-unlabeled-namespaces)
+  - [PodSecurityPolicy Migration](#podsecuritypolicy-migration)
+  - [Custom Profiles](#custom-profiles)
+  - [Custom Warning Messages](#custom-warning-messages)
+  - [Windows Support](#windows-support)
+  - [Offline Policy Checking](#offline-policy-checking)
+  - [Event recording](#event-recording)
+  - [Conformance](#conformance)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Replace PodSecurityPolicy with a new built-in admission controller that enforces the
+[Pod Security Standards].
+
+- Policy enforcement is controlled at the namespace level through labels
+- Policies can be applied in 3 modes. Multiple modes can apply to a single namespace.
+    - Enforcing: policy violations cause the pod to be rejected
+    - Audit: policy violations trigger an audit annotation, but are otherwise allowed
+    - Warning: policy violations trigger a user-facing warning, but are otherwise allowed
+- An optional per-mode version label can be used to pin the policy to the version that shipped
+  with a given Kubernetes minor version (e.g. `v1.18`)
+- Dry-run of namespace updates is supported to test enforcing policy changes against existing pods.
+- Policy exemptions can be statically configured based on (requesting) user, RuntimeClass, or
+  namespace. A request meeting exemption criteria is ignored by the admission plugin.
+
+[Pod Security Standards]: https://kubernetes.io/docs/concepts/security/pod-security-standards/
+
+## Motivation
+
+Pod Security Policy is deprecated as of Kubernetes v1.21. There were numerous problems with
+PSP that lead to the decision to deprecate it, rather than promote it to GA, including:
+
+1. Policy authorization model - Policies are bound to the requesting user OR the pod’s service
+   account
+     - Granting permission to the user is intuitive, but breaks controllers
+     - Dual model weakens security
```

Comment:
Binding PSP via service account gave an illusion of fine grained control, while the replacement might seem very coarse grained since same policy applies to all pods in namespace.  Would it be worth of explaining here, that the PSP model did not prevent users potentially working around fine-grained policy bindings by changing pod's service account, and that only namespace level policy could be enforced in practise?